Standard L2 Genesis

tynes opened the pull request:
**Description**

Moves network specific values into L1Block.sol.

Fixes: #12297, #12349, #12298, #12299, #12300, #12301, #12303, #12304, #12305, #12655 

## Open items: 

- [ ] Debug the failing [IL1BlockInterop interface](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/69881/workflows/88fb34f4-c48d-4400-b78f-5434743b86a5/jobs/2885955?invite=true#step-117-9) check. It does not behave the same on my machine (I have failures in both IL1Block and IL1BlockInterop). It's related to the constructor afaict, but I can't make it work.
- [x] Fix the one failing forge test, [`test_upgrade_correctEvent_succeeds`](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/69885/workflows/7a406985-ded7-449a-ab24-8d7e79a37c45/jobs/2885934?invite=true#step-112-750)
- [x] Fix the [unused imports check](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/69885/workflows/7a406985-ded7-449a-ab24-8d7e79a37c45/jobs/2885964?invite=true#step-118-87)
- [x] Add `feeAdmin` to op-deployer
- [ ] Consider if the new `SystemConfig.Roles` struct should also contain the `unsafeBlockSigner`
	- [ ] There is also a [`Roles` struct](https://github.com/ethereum-optimism/optimism/blob/feat/holocene-contracts/packages/contracts-bedrock/scripts/DeployOPChain.s.sol#L342) in `OPCM`. This is a bit confusing, can we disambiguate it?
- [x] Fix the semgrep findings

Cut from scope: 
- [x] #12655
	- [ ] Blocked by #12356

tynes committed:
wip: holocene contracts

tynes committed:
progress

tynes committed:
l1xdm

tynes commented on packages/contracts-bedrock/src/L2/L2ProxyAdmin.sol:
```diff
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
```
Consider refactoring the `ProxyAdmin` to remove the L1 specific features and make a `L1ProxyAdmin`

tynes committed:
standard bridge

tynes committed:
erc721bridge

tynes committed:
fee vaults

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -287,13 +287,8 @@ contract L2Genesis is Deployer {
-        vm.store(Predeploys.PROXY_ADMIN, _ownerSlot, bytes32(uint256(uint160(cfg.proxyAdminOwner()))));
         // update the proxy to not be uninitialized (although not standard initialize pattern)
-        vm.store(impl, _ownerSlot, bytes32(uint256(uint160(cfg.proxyAdminOwner()))));
+        vm.store(impl, bytes32(0), bytes32(uint256(0xdead)));
```
todo: double check if this is necessary, this adds complexity to the genesis definition

maurelian replied:
I can't think of why it would be necessary... is there something in storage at that slot on OP Mainnet?

tynes committed:
cleanup

tynes committed:
portal interop: cleanup

tynes committed:
cleanup

tynes committed:
contracts: fix erc721 factory tests

tynes committed:
contracts: more cleanup

tynes committed:
contracts: L2xdm tests passing

tynes committed:
contracts: l2 standard bridge tests

tynes committed:
tests: sequencer fee vault passes

tynes committed:
tests: sequencer fee vault passes

tynes committed:
tests: more passing

tynes committed:
contracts: portal tests pass

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -369,8 +366,7 @@ contract L2Genesis is Deployer {
     function setOptimismMintableERC721Factory() public {
-        OptimismMintableERC721Factory factory =
-            new OptimismMintableERC721Factory({ _bridge: Predeploys.L2_ERC721_BRIDGE, _remoteChainId: cfg.l1ChainID() });
+        OptimismMintableERC721Factory factory = new OptimismMintableERC721Factory();
```
This shouldn't be deploying the code directly with `new`

tynes commented on packages/contracts-bedrock/src/L1/OptimismPortal2.sol:
```diff
@@ -606,7 +608,31 @@ contract OptimismPortal2 is Initializable, ResourceMetering, ISemver {
+    ///         if (msg.sender != ISuperchainConfig.upgrader()) revert Unauthorized();
+    ///         this removes the need to have the system config know about the superchain config
+    function upgrade(address payable _proxy, address _implementation) external {
+        if (msg.sender != address(systemConfig)) revert Unauthorized();
```
This should be updated to expect a call directly and from the superchain config upgrader

maurelian replied:
https://github.com/ethereum-optimism/optimism/blob/feat/holocene-contracts/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L637

tynes commented on packages/contracts-bedrock/src/L2/L2ERC721Bridge.sol:
```diff
@@ -32,16 +33,24 @@ contract L2ERC721Bridge is ERC721Bridge, ISemver {
+    /// TODO: this should not be initializable, this should have messenger be abstract in the base
+    ///       only the L1 contract should be initializable
+    function initialize() public initializer {
+        __ERC721Bridge_init();
```
This needs to be removed

tynes commented on packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:
```diff
@@ -818,6 +818,20 @@ contract Deploy is Deployer {
     ////////////////////////////////////////////////////////////////
 
+    /// @notice Initialize the SuperchainConfig
+    function initializeSuperchainConfig() public broadcast {
```
This may be dead code now

maurelian replied:
Correct. It's no longer on `develop`.

tynes commented on packages/contracts-bedrock/src/L1/L1ERC721Bridge.sol:
```diff
@@ -42,7 +45,18 @@ contract L1ERC721Bridge is ERC721Bridge, ISemver {
     function initialize(ICrossDomainMessenger _messenger, ISuperchainConfig _superchainConfig) public initializer {
         superchainConfig = _superchainConfig;
-        __ERC721Bridge_init({ _messenger: _messenger, _otherBridge: ERC721Bridge(payable(Predeploys.L2_ERC721_BRIDGE)) });
+        crossDomainMessenger = _messenger;
```
This initialize needs to be removed

maurelian replied:
I'm not sure I follow. Are you saying that the `initialize()` function needs to be removed from this L1 contract? Or perhaps you are referring to `__ERC721Bridge_init` which already seems to have been removed?

tynes commented on packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol:
```diff
@@ -1,8 +1,9 @@
+pragma solidity 0.8.25;
 
 // Contracts
 import { CrossDomainMessenger } from "src/universal/CrossDomainMessenger.sol";
```
In the past, we did these storage layout spacers, oz v5 initializable uses unstructured storage, much simpler storage layout migration

tynes replied:
Using oz v5 requires solve ^0.8.20

maurelian replied:
We won't have #12356 in before this unfortunately.

tynes committed:
fix build

tynes committed:
messenger: fixup

tynes committed:
messenger fixup

tynes committed:
standard bridge: fixup

tynes committed:
fixes

tynes committed:
progress

tynes committed:
more tests passing

tynes committed:
fixups

tynes committed:
tests: cleanup

tynes committed:
more fixes

tynes committed:
more fixup

tynes committed:
more fixes

tynes committed:
more fixes

tynes committed:
tests: all passing

tynes committed:
tests

tynes commented:
Another possible feature to include is https://github.com/ethereum-optimism/optimism/issues/12136. This would make the enshrined revenue sharing predeploy much simpler to implement. This is a 2 line change and is backwards compatible, we could also wait to roll this out until we get around to shipping the enshrined revshare contract

maurelian committed:
fix: ERC721Bridge spacer offset

maurelian committed:
fix: lint

maurelian committed:
fix: Place comment above @custom:semver

This fixes the check which includes the comment in the parsed semver

maurelian committed:
chore: Update semver-lock

maurelian committed:
add some todos

maurelian commented on packages/contracts-bedrock/src/L1/OptimismPortal.sol:
```diff
@@ -572,17 +572,17 @@ contract OptimismPortal is Initializable, ResourceMetering, ISemver {
 
-    /// @notice Sets the gas paying token for the L2 system. This token is used as the
-    ///         L2 native asset. Only the SystemConfig contract can call this function.
-    function setGasPayingToken(address _token, uint8 _decimals, bytes32 _name, bytes32 _symbol) external {
```
@tynes did you mean to overwrite this function?

maurelian replied:
oh, nevermind, I see that it's a ConfigType now.

maurelian committed:
test: Add fuzz test for OptimismPortal2 setConfig

test: Add fuzz test for OptimismPortal2 setConfig succeeds

maurelian committed:
test: OptimismPortalInterop - test all auth failures in one fuzz test

test: OptimismPortalInterop - test all setConfig succeeds in one fuzz test

maurelian committed:
fix: Portal2 iface

maurelian commented on packages/contracts-bedrock/src/L1/OptimismPortal2.sol:
```diff
@@ -606,7 +608,30 @@ contract OptimismPortal2 is Initializable, ResourceMetering, ISemver {
+    /// @notice Calls the L2ProxyAdmin as the DEPOSITOR_ACCOUNT. This function can be used
+    ///         to upgrade the predeploys on L2. Only callable by the upgrader role on the
+    ///         SuperchainConfig.
+    function upgrade(uint32 _gasLimit, bytes memory _calldata) external {
```
@tynes any reason we don't just pass this in as `uint64` instead of casting it below?

tynes replied:
It could make more sense to use u64 given that the messenger uses u64 for its gas param. Better to be consistent

maurelian committed:
wip: Adding OptimismPortal2.upgrade() tests

maurelian committed:
feat: Set upgrader to superchainProxyAdminOwner in DeploySuperchain.s.sol

maurelian committed:
fix some ifaces

maurelian committed:
Some iface fixes

maurelian committed:
More iface fixes

maurelian committed:
fix solc warning

maurelian committed:
feat: Add setFeeVaultConfig

Replaces the 3 distinct functions with a single flexible function

maurelian committed:
wip adding Roles struct with feeadmin

maurelian committed:
Merge branch 'develop' into feat/holocene-contracts

maurelian committed:
update forge-std

very unsure why the merge from the develop didn't just do this.

maurelian committed:
feat: Add feeAdmin to new SystemConfig.Roles struct

maurelian committed:
feat: Fix DeployOPChain init assertions

maurelian committed:
feat: Update name to L1OptimismMintableERC20Factory

maurelian committed:
fix: testFuzz_upgrade

maurelian committed:
rename to L1OpMintableErc20Factory

maurelian committed:
fix: interace fixes

maurelian committed:
Merge branch 'develop' into feat/holocene-contracts

maurelian committed:
fix: delete unused functions

maurelian committed:
fix: spacer naming

maurelian committed:
fix some semvers

maurelian committed:
fix semvers

maurelian committed:
chore: gas benchmarks

maurelian committed:
fix spacer

maurelian commented:
@tynes I'm opening this up for your initial review. 

Status: 
1. Contract code should be feature complete. 
2. There is a single failing contract test I still need to debug.
3. Many other checks still need fixing (spacers, semgrep, interfaces, etc). 
4. Still want to migrate other L1 contracts to OZ' v5 Initializable
5. Still need to update `op-deployer` to support the new `feeAdmin` role

maurelian committed:
fix: interfaces

maurelian committed:
Remove outdated comment

maurelian committed:
fix: interfaces

maurelian committed:
fix: IL1Block

maurelian committed:
snapshots and semver-lock

maurelian committed:
fix: unused imports

maurelian committed:
fix: spacer and semver

maurelian committed:
fix: L120factory name

maurelian committed:
op-deployer: Add FeeAdmin

maurelian committed:
semgrep: More flexible string.concat usage

tynes commented on packages/contracts-bedrock/scripts/DeployImplementations.s.sol:
```diff
@@ -616,6 +614,9 @@ contract DeployImplementations is Script {
+            // TODO: rename?
+            // This is used to set the name of the contract in the
+            // `OPContractsManager.implementations.releases` mapping. Can that change between releases?
             name: "OptimismMintableERC20Factory",
```
I tried to rename this and it broke a bunch of stuff, didn't want to go down the whole rabbit hole of trying to fix it. I figure that the easiest way to handle a rename would be to allow for an OR in the naming for some time until we have upgraded all chains sufficiently past this point. It could also be more trouble than its worth doing the rename, the maintainers of OPCM should make this decision

maurelian replied:
I went through exactly the same experience, made the change then reverted it. Technically it should be OK to rename contracts between releases, but the number of things that broke was quite surprising so I decided just to revert it. It's fine as is for now.

tynes commented on packages/contracts-bedrock/scripts/DeployImplementations.s.sol:
```diff
@@ -400,22 +401,19 @@ contract DeployImplementationsOutput is BaseDeployIO {
         require(address(messenger.otherMessenger()) == Predeploys.L2_CROSS_DOMAIN_MESSENGER, "L1xDM-20");
         require(address(messenger.PORTAL()) == address(0), "L1xDM-30");
         require(address(messenger.portal()) == address(0), "L1xDM-40");
         require(address(messenger.superchainConfig()) == address(0), "L1xDM-50");
```
in the proxy you do the following check

```solidity
        vm.expectRevert("CrossDomainMessenger: xDomainMessageSender is not set");
        messenger.xDomainMessageSender();
```

This could also be useful here

maurelian replied:
`expectRevert` is [not yet supported in Go](https://github.com/ethereum-optimism/optimism/issues/12720). I will add a todo and some uncommented code in both cases instead.

tynes commented on packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:
```diff
@@ -823,7 +827,8 @@ contract Deploy is Deployer {
             _data: abi.encodeCall(
                 ISystemConfig.initialize,
                 (
```
I think it should be a new config value at this layer of abstraction and `op-deployer` can handle setting it to the correct thing. For new chains in the superchain, it should not be the same value as the owner of the system config but some chains may want it to be, so by adding a new config value it allows for flexibility while at a higher abstraction the `useStandard` config can make sure the correct value is set

tynes commented on packages/contracts-bedrock/scripts/libraries/ForgeArtifacts.sol:
```diff
@@ -214,7 +228,10 @@ library ForgeArtifacts {
-        bytes memory rawSlot = vm.parseJson(string(Process.run(command)));
+
+        bytes memory result = Process.run(command);
+        require(result.length > 0, string.concat("ForgeArtifacts:", _contractName, "is not initializable"));
```
Is this wrong because there is no space after the `:` ?

maurelian replied:
yes, fixed here: 156b1f915032

tynes commented on packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol:
```diff
@@ -57,7 +57,11 @@ contract L1CrossDomainMessenger is CrossDomainMessenger, ISemver {
-        __CrossDomainMessenger_init({ _otherMessenger: CrossDomainMessenger(Predeploys.L2_CROSS_DOMAIN_MESSENGER) });
+    }
+
+    /// @notice
```
needs natspec

maurelian replied:
7c507b4e8a9f

tynes commented on packages/contracts-bedrock/src/L1/L1ERC721Bridge.sol:
```diff
@@ -1,8 +1,9 @@
 // SPDX-License-Identifier: MIT
-pragma solidity 0.8.15;
+pragma solidity 0.8.25;
```
This was too painful to manage the storage layout diff without using v5 oz initializable

tynes commented on packages/contracts-bedrock/src/L1/L1ERC721Bridge.sol:
```diff
@@ -19,29 +20,42 @@ import { IL2ERC721Bridge } from "src/L2/interfaces/IL2ERC721Bridge.sol";
+        crossDomainMessenger = _messenger;
+    }
+
+    /// @notice
```
needs natspec

maurelian replied:
7c507b4e8a9f

tynes commented on packages/contracts-bedrock/src/L1/L1OptimismMintableERC20Factory.sol:
```diff
@@ -0,0 +1,27 @@
+        standardBridge = _bridge;
+    }
+
+    /// @notice
```
Needs natspec

maurelian replied:
7c507b4e8a9f

tynes commented on packages/contracts-bedrock/src/L1/L1StandardBridge.sol:
```diff
@@ -106,10 +110,17 @@ contract L1StandardBridge is StandardBridge, ISemver {
+        return StandardBridge(payable(Predeploys.L2_STANDARD_BRIDGE));
+    }
+
+    // TODO:
```
needs natspec

maurelian replied:
7c507b4e8a9f

tynes commented on packages/contracts-bedrock/src/L1/SuperchainConfig.sol:
```diff
@@ -36,19 +42,20 @@ contract SuperchainConfig is Initializable, ISemver {
     /// @notice Constructs the SuperchainConfig contract.
     constructor() {
-        initialize({ _guardian: address(0), _paused: false });
+        initialize({ _guardian: address(0), _upgrader: address(0), _paused: false });
```
It could make sense to use `_disableInitializers` here

tynes commented on packages/contracts-bedrock/src/L1/SystemConfig.sol:
```diff
@@ -4,17 +4,26 @@ pragma solidity 0.8.15;
+import { IOptimismPortal2 as IOptimismPortal } from "src/L1/interfaces/IOptimismPortal2.sol";
 import { IResourceMetering } from "src/L1/interfaces/IResourceMetering.sol";
 
+// TODO: need rbac construction for being able to set the fee vault config to keep
```
This todo comment can be deleted

tynes commented on packages/contracts-bedrock/src/L1/SystemConfig.sol:
```diff
@@ -48,6 +57,12 @@ contract SystemConfig is OwnableUpgradeable, ISemver, IGasToken {
+        address owner;
+        address feeAdmin;
+    }
+    // TODO: add unsafe block signer and batcher hash?
```
It could make sense to do this, we could also do batcher as address rather than bytes32 to improve the ux. Also could be good to rename owner to chainOperator or chainGovernor

tynes commented on packages/contracts-bedrock/src/L2/FeeVault.sol:
```diff
@@ -15,27 +16,14 @@ import { Types } from "src/libraries/Types.sol";
     /// @notice The minimum gas limit for the FeeVault withdrawal transaction.
     uint32 internal constant WITHDRAWAL_MIN_GAS = 400_000;
 
+    /// @notice
```
natspec

maurelian replied:
7c507b4e8a9f

tynes commented on packages/contracts-bedrock/src/L2/L1Block.sol:
```diff
@@ -168,14 +197,100 @@ contract L1Block is ISemver, IGasToken {
+    }
+
+    /// @notice Helper function for migrating deploy config.
+    ///         TODO: may need to make 3 calls for legacy purposes, the config abi doesn't yet exist
```
This is important, we cannot use the `config()` function as it will not exist when this migration is happening

tynes commented on packages/contracts-bedrock/src/L2/L1Block.sol:
```diff
@@ -168,14 +197,100 @@ contract L1Block is ISemver, IGasToken {
+    /// @notice
+    function setHolocene() external {
+        if (msg.sender != DEPOSITOR_ACCOUNT()) revert NotDepositor();
+        // TODO set holocene activation to true
```
Similar to the `GasPriceOracle`, we should consider an `isHolocene()(bool)` getter that is set to `true` in this call

tynes commented on packages/contracts-bedrock/src/L2/L2ERC721Bridge.sol:
```diff
@@ -26,21 +28,19 @@ import { ISemver } from "src/universal/interfaces/ISemver.sol";
-    /// @notice Constructs the L2ERC721Bridge contract.
-    constructor() ERC721Bridge() {
-        initialize({ _l1ERC721Bridge: payable(address(0)) });
+    /// @notice
```
natspec

tynes commented on packages/contracts-bedrock/src/L2/OptimismMintableERC721Factory.sol:
```diff
@@ -2,19 +2,19 @@
-    /// @notice Chain ID for the remote network.
-    uint256 public immutable REMOTE_CHAIN_ID;
-
+    // TODO: check storage layout
```
dead todo

tynes commented on packages/contracts-bedrock/src/L2/OptimismMintableERC721Factory.sol:
```diff
@@ -25,27 +25,31 @@ contract OptimismMintableERC721Factory is ISemver {
-    /// @notice Address of the ERC721 bridge on this network.
-    function bridge() external view returns (address) {
-        return BRIDGE;
+    /// @notice TODO: type should be more strict
```
natspec

maurelian replied:
7c507b4e8a9f

tynes commented on packages/contracts-bedrock/src/libraries/StaticConfig.sol:
```diff
@@ -57,4 +64,46 @@ library StaticConfig {
+    }
+
+    /// @notice
+    function decodeSetFeeVaultConfig(bytes memory _data)
```
This is dead code

tynes commented on packages/contracts-bedrock/src/libraries/StaticConfig.sol:
```diff
@@ -1,6 +1,13 @@
 
+import { Types } from "src/libraries/Types.sol";
+
+// The main benefit of this library is to give the reader the ability to observe that
```
we should delete this comment

tynes commented on packages/contracts-bedrock/src/libraries/StaticConfig.sol:
```diff
@@ -1,6 +1,13 @@
+
 /// @title StaticConfig
 /// @notice Library for encoding and decoding static configuration data.
 library StaticConfig {
```
i have a preference for deleting this whole library, its verbose

tynes commented on packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol:
```diff
@@ -308,24 +311,36 @@ abstract contract CrossDomainMessenger is
+        return xDomainMsgSender;
+    }
+
+    /// @notice
```
natspec

maurelian replied:
7c507b4e8a9f

tynes commented on packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol:
```diff
@@ -308,24 +311,36 @@ abstract contract CrossDomainMessenger is
     }
 
+    /// @notice
+    function getCrossDomainMessageSender() internal view returns (address) {
```
this could also be transient storage

maurelian replied:
We unfortunately won't have the solidity version update in yet.

maurelian committed:
update todo notes

maurelian committed:
fix: Semgrep rules and source formatting

maurelian committed:
Remove unused storage and immutable checks from TestApplyGenesisStrategy

Chain specific intent data should no longer appear in the L2 Genesis,
as it will be the same for all chains.

maurelian committed:
fix: test_upgrade_correctEvent_succeeds bug

maurelian committed:
remove unused console import

maurelian committed:
fix: Specs for SystemConfigInterop abi

maurelian committed:
fix: semver diffs on required contracts

maurelian committed:
remove unused getEIP1967ImplementationAddress func

maurelian committed:
fix: layout in StandardBridge

maurelian committed:
fix: go lint

maurelian committed:
fix: layout in standard bridge

maurelian committed:
feat: use _disableInitializers on L1StandardBridge and SuperchainConfig

maurelian committed:
fix: L1xDM spacers

maurelian commented on op-deployer/pkg/deployer/integration_test/apply_test.go:
```diff
@@ -365,64 +360,19 @@ func validateOPChainDeployment(t *testing.T, cg codeGetter, st *state.State, int
-		checkStorageSlot(t, alloc, predeploys.ProxyAdminAddr, ownerSlot, addrAsSlot)
 		var defaultGovOwner common.Hash
 		defaultGovOwner.SetBytes(common.HexToAddress("0xDeaDDEaDDeAdDeAdDEAdDEaddeAddEAdDEAdDEad").Bytes())
-		checkStorageSlot(t, alloc, predeploys.GovernanceTokenAddr, common.Hash{31: 0x0a}, defaultGovOwner)
```
Restore this check.

maurelian commented on op-deployer/pkg/deployer/integration_test/apply_test.go:
```diff
@@ -365,64 +360,19 @@ func validateOPChainDeployment(t *testing.T, cg codeGetter, st *state.State, int
 		addrAsSlot.SetBytes(chainIntent.Roles.L1ProxyAdminOwner.Bytes())
-		// slot 0
-		ownerSlot := common.Hash{}
-		checkStorageSlot(t, alloc, predeploys.ProxyAdminAddr, ownerSlot, addrAsSlot)
```
This check no longer applies since the L2ProxyAdmin owner is [hardcoded](https://github.com/ethereum-optimism/optimism/blob/feat/holocene-contracts/packages/contracts-bedrock/src/L2/L2ProxyAdmin.sol#L15).

maurelian commented on packages/contracts-bedrock/scripts/DeployOPChain.s.sol:
```diff
@@ -566,8 +567,9 @@ contract DeployOPChain is Script {
-        require(address(uint160(uint256(xdmSenderSlot))) == Constants.DEFAULT_L2_SENDER, "L1xDM-60");
+        // TODO: vm.expectRevert is not supported by op-chain-ops, so we can't check this yet.
+        // vm.expectRevert("CrossDomainMessenger: xDomainMessageSender is not set");
+        // messenger.xDomainMessageSender();
```
Add a link to the [issue here](https://github.com/ethereum-optimism/optimism/issues/12720).

maurelian commented on packages/contracts-bedrock/scripts/DeployOPChain.s.sol:
```diff
@@ -345,8 +345,9 @@ contract DeployOPChain is Script {
-            challenger: _doi.challenger()
-        });
+            challenger: _doi.challenger(),
+            feeAdmin: msg.sender // TODO: read from doi
```
Note this todo.

maurelian commented on packages/contracts-bedrock/scripts/libraries/ForgeArtifacts.sol:
```diff
@@ -214,7 +228,10 @@ library ForgeArtifacts {
-        bytes memory rawSlot = vm.parseJson(string(Process.run(command)));
+
+        bytes memory result = Process.run(command);
+        require(result.length > 0, string.concat("ForgeArtifacts: ", _contractName, "is not initializable"));
```
This should no longer be an issue. 
Fix here: 156b1f915032
Clean CI run: https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/70354/workflows/b54fb881-c4c0-44e5-902c-045a0913ee69/jobs/2901864

maurelian commented on packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol:
```diff
@@ -119,11 +112,12 @@ abstract contract CrossDomainMessenger is
+    /// @notice Spacer for backwards compatibility. The storage slot was migrated when the
+    ///         initializer pattern was moved away from in the base contract to remove the
+    ///         need to set `Constants.DEFAULT_L2_SENDER` into storage during a call to `initialize`.
+    bytes20 internal spacer_204_0_20;
```
it's not clear to me why this needed to move. The initialized slot comes afterwards.

maurelian replied:
I grok the comment now, and also that this removes the need to zero out the existing value in the L2 contract.

maurelian commented on packages/contracts-bedrock/snapshots/storageLayout/L1CrossDomainMessenger.json:
```diff
@@ -113,17 +106,52 @@
+    "label": "xDomainMsgSender",
     "offset": 0,
     "slot": "208",
-    "type": "uint256[43]"
```
This gap is now 41 slots long, and is shifted down. 
It allows for the new `xDomainMsgSender` and `initialized` slots to each occupy a full 32 bytes slot, at the beginning of where this gap was.

The key thing to note is that in this layout is that the pre-existing storage slots are either:
1. unaffected
2. have been replaced by a spacer

maurelian commented on packages/contracts-bedrock/snapshots/storageLayout/L1StandardBridge.json:
```diff
@@ -36,37 +22,58 @@
-    "offset": 0,
-    "slot": "50",
+    "offset": 2,
+    "slot": "49",
```
Need a spacer above to prevent packing the `superchainConfig` into the `_initialized` slot.

maurelian commented on packages/contracts-bedrock/snapshots/storageLayout/L2CrossDomainMessenger.json:
```diff
@@ -113,16 +106,30 @@
+    "label": "__gap",
+    "offset": 0,
+    "slot": "209",
+    "type": "uint256[41]"
```
The gap size is reduced by 2, but should be reduced by 1 to maintain parity with the previous layout.

maurelian commented on packages/contracts-bedrock/snapshots/storageLayout/L2OptimismMintableERC20Factory.json:
```diff
@@ -0,0 +1,30 @@
+    "slot": "3",
+    "type": "uint256[48]"
+  }
+]
```
This is net new, but needs to be compared to the previous `OptimismMintableERC20Factory.json`:

```
git checkout develop snapshots/storageLayout/OptimismMintableERC20Factory.json
diff snapshots/storageLayout/OptimismMintableERC20Factory.json snapshots/storageLayout/L2OptimismMintableERC20Factory.json
```


We can see that `_initialized` and `bridge` are removed, and `deployments` is unaffected.
![2024-10-30 at 09 51@2x](https://github.com/user-attachments/assets/44d76237-d75e-4442-80e5-0657e7c64ea4)

maurelian commented on packages/contracts-bedrock/snapshots/storageLayout/L2ProxyAdmin.json:
Comparing this to the base ProxyAdmin, has empty output:
```
diff snapshots/storageLayout/ProxyAdmin.json snapshots/storageLayout/L2ProxyAdmin.json
```

maurelian commented on packages/contracts-bedrock/snapshots/storageLayout/L2StandardBridgeInterop.json:
```diff
@@ -36,23 +22,23 @@
     "offset": 0,
     "slot": "5",
-    "type": "uint256[45]"
+    "type": "uint256[44]"
```
I don't think this should have changed.

maurelian commented on packages/contracts-bedrock/snapshots/storageLayout/L2StandardBridge.json:
```diff
@@ -1,24 +1,10 @@
-    "offset": 2,
-    "slot": "0",
-    "type": "bytes30"
+    "type": "bytes32"
```
Spacer size increased to cover `_initialized`, subsequent slots are unaffected.

maurelian commented on packages/contracts-bedrock/snapshots/storageLayout/L2StandardBridge.json:
```diff
@@ -36,23 +22,23 @@
     "offset": 0,
     "slot": "3",
-    "type": "contract ICrossDomainMessenger"
+    "type": "address"
```
moved to a constant

maurelian commented on packages/contracts-bedrock/snapshots/storageLayout/L2StandardBridge.json:
```diff
@@ -36,23 +22,23 @@
     "offset": 0,
     "slot": "4",
-    "type": "contract StandardBridge"
+    "type": "address"
```
moved to L1Block

maurelian commented on packages/contracts-bedrock/snapshots/storageLayout/L2StandardBridge.json:
```diff
@@ -36,23 +22,23 @@
     "offset": 0,
     "slot": "5",
-    "type": "uint256[45]"
+    "type": "uint256[44]"
```
no need to change this length.

maurelian commented on packages/contracts-bedrock/snapshots/storageLayout/L2StandardBridgeInterop.json:
The same comments left on `packages/contracts-bedrock/snapshots/storageLayout/L2StandardBridge.json` (immediately above) also apply here.

maurelian committed:
feat: Add missing natspec

maurelian committed:
feat: Add missing natspec

maurelian committed:
fix: Spacer to pad out L1StandardBridge _initialized slot

maurelian committed:
feat: Make base CDM abstract, remove its snapshots

maurelian committed:
fix: spacer sizing on L1CDM

maurelian committed:
fix: L1StandardBridge init check

maurelian committed:
fix: spacer validations

tynes commented:
Going to close this PR due to it being replaced by another
